lib/actions-oclif/tunnel.ts

```diff
@@ -0,0 +1,245 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { flags } from '@oclif/command';
+import Command from '../command';
+import {
+	NoPortsDefinedError,
+	InvalidPortMappingError,
+	ExpectedError,
+} from '../errors';
+import * as cf from '../utils/common-flags';
+import { getBalenaSdk, stripIndent } from '../utils/lazy';
+import { getOnlineTargetUuid } from '../utils/patterns';
+import * as _ from 'lodash';
+import { tunnelConnectionToDevice } from '../utils/tunnel';
+import { createServer, Server, Socket } from 'net';
+import * as Bluebird from 'bluebird';
+import { tryAsInteger } from '../utils/validation';
+import { IArg } from '@oclif/parser/lib/args';
+
+interface FlagsDef {
+	port: string[];
+	help: void;
+}
+
+interface ArgsDef {
+	deviceOrApplication: string;
+}
+
+export default class TunnelCmd extends Command {
+	public static description = stripIndent`
+		Tunnel local ports to your balenaOS device.
+
+		Use this command to open local ports which tunnel to listening ports on your balenaOS device.
+
+		For example, you could open port 8080 on your local machine to connect to your managed balenaOS
+		device running a web server listening on port 3000.
+
+		Port mappings are specified in the format: <remotePort>[:[localIP:]localPort]
+		localIP defaults to 'localhost', and localPort defaults to the specified remotePort value.
+
+		You can tunnel multiple ports at any given time.
+	`;
+
+	public static examples = [
+		'# map remote port 22222 to localhost:22222',
+		'$ balena tunnel abcde12345 -p 22222',
+		'',
+		'# map remote port 22222 to localhost:222',
+		'$ balena tunnel abcde12345 -p 22222:222',
+		'',
+		'# map remote port 22222 to any address on your host machine, port 22222',
+		'$ balena tunnel abcde12345 -p 22222:0.0.0.0',
+		'',
+		'# map remote port 22222 to any address on your host machine, port 222',
+		'$ balena tunnel abcde12345 -p 22222:0.0.0.0:222',
+		'',
+		'# multiple port tunnels can be specified at any one time',
+		'$ balena tunnel abcde12345 -p 8080:3000 -p 8081:9000',
+	];
+
+	public static args: Array<IArg<any>> = [
+		{
+			name: 'deviceOrApplication',
+			description: 'device uuid or application name/id',
+			parse: (x) => tryAsInteger(x),
+			required: true,
+		},
+	];
+
+	public static usage = 'tunnel <deviceOrApplication>';
+
+	public static flags: flags.Input<FlagsDef> = {
+		port: flags.string({
+			description:
+				'port mapping in the format <remotePort>[:[localIP:]localPort]',
+			char: 'p',
+			multiple: true,
```

Comment:
Yeah it works but it has caveats.   It's supposed to work like `-p one -p two -p three` (and it does), but it also tries to process like `-p one two three`, and gets mixed up then when there are args after flags.  There's community PR submitted which addresses this, but it hasn't been acknowledged by the oclif team for over a month (despite a couple of pings, including one from me).  As long as the users stick to the format in the examples, it will be ok.